dali/imgcodec/decoders/nvjpeg2k/nvjpeg2k.cc

```diff
@@ -237,31 +173,47 @@ DecodeResult NvJpeg2000DecoderInstance::DecodeImplTask(int thread_idx,
   if (!ParseJpeg2000Info(in, ctx))
     return result;
 
-  const int64_t channels = ctx.shape[0];
+  const int64_t channels = ctx.image_info.num_components;
+  bool is_roi_oob = ctx.roi &&
+    (ctx.roi.begin[0] < 0 || ctx.roi.end[0] > ctx.image_info.image_width ||
+     ctx.roi.begin[1] < 0 || ctx.roi.end[1] > ctx.image_info.image_height);
+
   DALIImageType format = channels == 1 ? DALI_GRAY : DALI_RGB;
   bool is_processing_needed =
     channels > 1 ||  // nvJPEG2000 decodes into planar layout
     ctx.pixel_type != opts.dtype ||
     format != opts.format ||
-    (ctx.bpp != 8 && ctx.bpp != 16);
+    (ctx.bpp != 8 && ctx.bpp != 16) ||
+    is_roi_oob;
 
   auto decode_out = out;
+  auto decode_sh = ctx.orig_shape;
+  ROI roi_decode;  // ROI passed to decode API
+  ROI roi_convert;  // ROI passed to convert API
+  if (ctx.roi && !is_roi_oob) {
+    // if ROI decoding and ROI withing bounds
+    // then use ROI decoding API directly
+    decode_sh = ctx.shape;
+    roi_decode = ctx.roi;
+  } else {
+    // Otherwise, we decode full image, then slice/pad via convert
+    decode_sh = ctx.orig_shape;
+    roi_convert =  ctx.roi;
+  }
+
+  kernels::DynamicScratchpad scratchpad({}, ctx.cuda_stream);
   if (is_processing_needed) {
     int64_t type_size = dali::TypeTable::GetTypeInfo(ctx.pixel_type).size();
-    size_t new_size = volume(ctx.shape) * type_size;
-    if (new_size > res.intermediate_buffer.capacity()) {
-      CUDA_CALL(cudaStreamSynchronize(ctx.cuda_stream));
-    }
-    res.intermediate_buffer.resize(volume(ctx.shape) * type_size);
-    decode_out = {res.intermediate_buffer.data(), ctx.shape, ctx.pixel_type};
+    size_t nbytes = volume(decode_sh) * type_size;
+    void *tmp_buff = scratchpad.AllocateGPU<uint8_t>(nbytes);
```

Comment:
Are we fine with alignment issues for `uint16_t` here (`.AllocateGPU<uint8_t>`)?